apps/editing-toolkit/editing-toolkit-plugin/coming-soon/coming-soon.php

```diff
@@ -31,6 +35,42 @@ function should_show_coming_soon_page() {
 	return apply_filters( 'a8c_show_coming_soon_page', $should_show );
 }
 
+/**
+ * Determines whether the coming soon page should be bypassed.
+ *
+ * It checks if share code is provided as GET parameter, or as a cookie.
+ * Then it validates the code against blog option, and if sharing code is valid,
+ * it allows bypassing the Coming Soon page.
+ *
+ * Finally, it sets a code in cookie and sets header that prevents robots from indexing.
+ *
+ * @return bool
+ */
+function is_accessed_by_valid_share_link() {
+	if ( isset( $_GET['share'] ) ) { // phpcs:ignore WordPress.Security.NonceVerification.Recommended
+		$share_code = $_GET['share']; // phpcs:ignore WordPress.Security.NonceVerification.Recommended
+	} elseif ( isset( $_COOKIE['share_code'] ) ) {
+		$share_code = $_COOKIE['share_code'];
+	} else {
+		$share_code = '';
+	}
+
+	$preview_links_option = get_option( 'wpcom_public_preview_links' );
+
+	if ( ! is_array( $preview_links_option ) || ! count( $preview_links_option ) || ! $share_code ) {
+		return false;
+	}
+
+	if ( ! in_array( $share_code, array_column( $preview_links_option, 'code' ), true ) ) {
+		return false;
+	}
+
+	setcookie( 'share_code', $share_code, time() + 3600, '/', false, is_ssl() );
+	header( 'X-Robots-Tag: noindex, nofollow' );
```

Comment:
The `wp_headers` hook is probably a better place to apply this header ([ref](https://developer.wordpress.org/reference/hooks/wp_headers/))

Reply:
Thanks. I tried to implement it, but the `wp_headers` filter is applied before the Coming Soon logic runs. 

I've checked the codebase and I couldn't find many usages of `wp_headers`, and I saw we use `header()` widely. Do you think it's worth digging into this more?

Reply:
@wojtekn I think `wp_headers` is best, yes. It's the most reliable way to safely add your own headers, without running into the risk of `Headers already sent`.

You can filter `wp_headers` and add our headers if the cookie or query param are present.

Reply:
@danielbachhuber ok, let's do this.